platforms/org.eclipse.mita.platform.xdk110/src/org/eclipse/mita/platform/xdk110/platform/Validation.xtend

```diff
@@ -149,28 +133,72 @@ class Validation implements IResourceValidator {
-		val i2cWrites  = new HashSet(i2cs ) => [retainAll(writes)]
-		val gpioReads  = new HashSet(gpios) => [retainAll(reads)]
-		val gpioWrites = new HashSet(gpios) => [retainAll(writes)]
-		
+		val i2cReads   = new HashSet(i2cs  ) => [retainAll(reads)]
+		val i2cWrites  = new HashSet(i2cs  ) => [retainAll(writes)]
+		val gpioReads  = new HashSet(gpios ) => [retainAll(reads)]
+		val gpioWrites = new HashSet(gpios ) => [retainAll(writes)]
+				
 		i2cReads.forEach[validateI2cReadWrite(it.source, it.sigInst, it.structFeature, "Read", "read from", acceptor)]
 		i2cWrites.forEach[validateI2cReadWrite(it.source, it.sigInst, it.structFeature, "Write", "write to", acceptor)]
 		
 		i2cWrites.forEach[validateI2cWriteLength(it.source, it.method, it.sigInst, acceptor)]
 		
 		gpioReads.forEach[validateGpioRead(it, acceptor)]
 		gpioWrites.forEach[validateGpioWrite(it, acceptor)]
+		
+		noises.forEach[validateNoiseRead(noises, it, acceptor)]
+	}
+	
+	/*
+	 * We validate if in timed events the noise sensor is only sampled as fast as it has samples here. 
+	 * Since this is equivalent to the halting problem we only emit warnings, not errors (for example in conflicting control flow) <-> we assume some kind of worst case: every read happens at least once.
+	 * Right now this is a bad estimate for if/else, where it is too pessimistic, and for loops, where it might be too optimistic.
+	 * I've also decided to not do this kind of validation in non-time event handlers like every button_one.pressed
+	 * since there the real constraints are timeout (which is handled by org.eclipse.mita.platform.xdk110.sensors.NoiseSensorValidator) 
+	 * and the real world (which we don't know anything about). 
+	 */
+	def validateNoiseRead(Set<MethodCallModality> noiseReads, MethodCallModality mcm, ValidationMessageAcceptor acceptor) {
+		val setups = mcm.source.eResource.resourceSet.allContents.filter(Program).flatMap[it.setup.iterator].filter[it.type?.name == NoiseSensorValidator.NOISE_SENSOR_TYPE_NAME];
+		val sampleTime_expr = NoiseSensorValidator.getTimePerSampleInMs(setups.head);
+		
+		var outerProgramBlock = EcoreUtil2.getContainerOfType(mcm.source, ProgramBlock);
+		var outerMost = false;
+		while(!outerMost) {
+			val next = EcoreUtil2.getContainerOfType(outerProgramBlock.eContainer, ProgramBlock);
+			if(next === null) {
+				outerMost = true;
+			}
+			else {
+				outerProgramBlock = next;
+			}
+		}
+		
+		val nearbyReads = outerProgramBlock.eAllContents.filter[obj | noiseReads.findFirst[it.source == obj] !== null].toList;
+		val previousNoiseReads = nearbyReads.takeWhile[it !== mcm.source].toList;
+		
+		val eventHandler = EcoreUtil2.getContainerOfType(mcm.source, EventHandlerDeclaration);
+		val event = eventHandler.event;
+		val numberOfPreviousReadsInThisEventHandler = previousNoiseReads.length + 1;
+		val actualSampleTime = sampleTime_expr.key * numberOfPreviousReadsInThisEventHandler;
+		if(event instanceof TimeIntervalEvent) {
```

Comment:
This is really cool. Maybe some upper/lower bounds would be handy (i.e. how fast can one sample noise sensor with the XDK "reliably"?)